File: services/data-pipeline-py/app/schemas/market.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class PriceTick(BaseModel):
    model_config = ConfigDict(extra="forbid")

    symbol: str
    price: float
    source: str
    timestamp: datetime
    scenario: str = "baseline"


class PublishQuotesRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    scenario: str = "baseline"
    symbols: list[str] = []
    transport: str = "http"


class ReplayQuotesRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    scenario: str = "baseline"
    symbols: list[str] = []
    speed_multiplier: float = Field(default=0.0, ge=0.0, le=1000.0)
    transport: str = "http"


class LiquidityCurvePoint(BaseModel):
    model_config = ConfigDict(extra="forbid")

    max_notional: float = Field(ge=0.0)
    fill_ratio: float = Field(ge=0.0, le=1.0)


class MarketMicrostructureProfile(BaseModel):
    model_config = ConfigDict(extra="forbid")

    signal_latency_ticks: int = Field(default=0, ge=0)
    spread_bps: float = Field(default=0.0, ge=0.0)
    max_fill_notional_per_tick: float = Field(default=0.0, ge=0.0)
    liquidity_curve: list[LiquidityCurvePoint] = Field(default_factory=list)
    queue_priority: float = Field(default=0.0, ge=0.0, le=1.0)
    market_impact_bps_per_10k: float = Field(default=0.0, ge=0.0)
    cancel_after_ticks: int = Field(default=0, ge=0)


class ScenarioCatalogEntry(BaseModel):
    model_config = ConfigDict(extra="forbid")

    scenario_id: str
    name: str
    description: str
    symbols: list[str]
    tick_count: int
    started_at: datetime
    ended_at: datetime
    tags: list[str]
    microstructure_profile: MarketMicrostructureProfile

File: services/simulator-ui/src/shared/types/market.ts
import type { IsoDateTimeString, QueryParams } from "./api";

export interface PriceTick {
  symbol: string;
  price: number;
  source: string;
  timestamp: IsoDateTimeString;
  scenario?: string;
}

export interface LatestQuoteQuery extends QueryParams {
  symbol: string;
  scenario?: string;
}

export interface PublishQuotesRequest {
  scenario?: string;
  symbols?: string[];
  transport?: string;
}

export interface ReplayQuotesRequest {
  scenario?: string;
  symbols?: string[];
  speed_multiplier: number;
  transport?: string;
}

export interface LatestQuoteResponse {
  quote: PriceTick;
}

export interface MarketScenariosResponse {
  scenarios: string[];
}

export interface ScenarioCatalogEntry {
  scenario_id: string;
  name: string;
  description: string;
  symbols: string[];
  tick_count: number;
  started_at: IsoDateTimeString;
  ended_at: IsoDateTimeString;
  tags: string[];
  microstructure_profile: MarketMicrostructureProfile;
}

export interface MarketMicrostructureProfile {
  signal_latency_ticks: number;
  spread_bps: number;
  max_fill_notional_per_tick: number;
  liquidity_curve?: LiquidityCurvePoint[];
  queue_priority?: number;
  market_impact_bps_per_10k?: number;
  cancel_after_ticks?: number;
}

export interface LiquidityCurvePoint {
  max_notional: number;
  fill_ratio: number;
}

export interface ScenarioCatalogResponse {
  scenarios: ScenarioCatalogEntry[];
}

export interface ScenarioDetailResponse {
  scenario: ScenarioCatalogEntry;
}

export interface PublishQuotesResponse {
  scenario: string;
  published_count: number;
  symbols: string[];
  transport: string;
  transports: string[];
}

export interface ReplayQuotesResponse {
  scenario: string;
  published_count: number;
  symbols: string[];
  speed_multiplier: number;
  transport: string;
}
